#  Multivariate Regression(GD)

In this notebook we are going to implement multivariate regression(gradient descent version). In particular, you will have to:

* Complete the function `cost_function` to implement cost function for multivariate regression(gradient descent version) algorithm.
* Complete the function `GDmultiLinparamEstimates` to implement multivariate regression(gradient descent version) algorithm.

Note we do not cover single value linear regression (gradient descent version) in this experiment as it is a similar one. You can play with it yourselves after going through this notebook.

# Import libraries

The required libraries for this notebook are pandas, sklearn and numpy.

In [2]:
# import libraries
import pandas
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# Load the data
The data we are using is from ***multi_regr_data.csv***. It consists of 1000 data related to student marks. Each data point has 3 columns(marks) and we are going to use all of them for multivariate linear regression. In particular, we will use the first 2 marks to predict the 3rd mark.


In [3]:
# Loading the CSV file
dataset=pandas.read_csv('./datasets/multi_regr_data.csv')
print(dataset.shape) #(data_number,feature_number)

(1000, 3)


# Split data into training and testing

In [4]:
# Split the data, we will use first 2 columns as features and the 3rd columns as target.
X = dataset[list(dataset.columns)[:-1]]
Y = dataset[list(dataset.columns)[-1]] 

# As pointed out in previous lab, we need to add a constant feature
intercept=np.ones((X.shape[0],1))
X=np.concatenate((intercept,X),axis=1)

# Split the data into training and testing(75% training and 25% testing data)
xtrain,xtest,ytrain,ytest=train_test_split(X, Y, random_state=0)


# Optimize using Gradient Descent Algorithm

The cost function is defined as follows:
\begin{align}
J\left(\beta \right) & =  {\frac{1}{2n}}\sum_{i=1}^n \left(y_i - \hat{y_i} \right)^2\\
\end{align}
or 
\begin{align}
J\left(\beta \right) & =  {\frac{1}{2n}}SSR\left(y_i,\hat{y_i} \right)\end{align}

You are asked to implement this cost function.

In [7]:
def cost_function(X, Y, beta):
    #  complete the code below.
    cost = np.square(Y - np.matmul(X, beta)).mean()/2
    
    return cost

# initialize B
beta= np.zeros(3, dtype=float)

# cal initial cost
inital_cost = cost_function(X, Y, beta)
print(inital_cost)

2470.11


**Gradient Descent Steps:**
1. Initialize values: \begin{align}\beta_0,\beta_1,…,\beta_n\end{align}  It is suggested you initilize with 0.
2. Iteratively update, until convergence: \begin{align} β_j : =  β_j - \alpha {\frac{\partial}{\partial J\left(\beta_j \right) }}J\left(\beta \right) \\ \end{align}  α: learning rate.

**Hint:** Step 2 function can also be written as \begin{align} β_j : =  β_j - \alpha \frac{1}{n}\sum_{i=1}^n \left(\hat{y_i} - y_i \right)x_{ij}\\ \end{align}

In [12]:

def GDmultiLinparamEstimates(X, Y, beta, learning_rate, iterations):
    cost_history = [0] * iterations
    n = X.shape[0]
    beta = np.zeros(X.shape[1], dtype=float)
    
    for iteration in range(iterations):
        #  complete the code below.   
        
        # Gradient Calculation
        gradient = np.dot((np.matmul(X, beta) - Y), X)/n
        # Changing Values of beta using Gradient
        beta = beta - learning_rate*gradient
        
        cost = cost_function(X, Y, beta)
        cost_history[iteration] = cost
        
    return beta, cost_history

iterations = 100000 # a value of iteration
learning_rate = 0.0001 # alpha



# run your algorithm
newB, cost_history = GDmultiLinparamEstimates(X, Y, beta, learning_rate, iterations)

# New Values of B
print(newB)
# Final Cost of new B
print(cost_history[-1])


[-0.47889172  0.09137252  0.90144884]
10.475123473539167


The final hypothesis for the whole dataset (i.e. X and Y as defined above) should be:
\begin{align}
y & = -0.47889172 + 0.09137252*x_1 + 0.90144884*x_2
\end{align}

# Congrat! 

Now, you have learned how to implement multivariate regression algorithm using gradient descent method. The next step would be to test your algorithm on test dataset, which we encourage you to do it yourself given the experience in previous lab.

In [13]:
newB, cost_history = GDmultiLinparamEstimates(xtrain, ytrain, beta, learning_rate, iterations)


In [37]:
from sklearn.metrics import accuracy_score
y_pred = np.dot(xtest,newB)
print(y_pred)

float64
